src/pass_first.rs

```diff
@@ -1,30 +1,58 @@
-use rio_api::{model::Triple, parser::TriplesParser};
+use rayon::prelude::*;
+use rio_api::parser::TriplesParser;
 use rio_turtle::TurtleError;
-use std::{
-    io::{stdin, BufRead, BufReader, Write},
-    path::Path,
-};
+use std::{io::Write, path::Path, sync::Mutex};
 
-use crate::io;
+use crate::{
+    io,
+    rdf_types::{Triple, TripleView},
+};
 
-fn index_triple(t: Triple, out: &mut impl Write) -> Result<(), TurtleError> {
-    match t.predicate.iri {
-        "http://www.w3.org/1999/02/22-rdf-syntax-ns#type" => {
-            let _ = out.write(&format!("{} .\n", &t.to_string()).into_bytes());
+fn index_triple(t: Triple, out: &mut impl Write) {
+    if t.predicate.iri.as_str() == "http://www.w3.org/1999/02/22-rdf-syntax-ns#type" {
+        if let Err(e) = out.write(&format!("{} .\n", &t.to_string()).into_bytes()) {
+            panic!("Error writting to out buffer: {e}");
         }
-        _ => {}
     }
-
-    Ok(())
 }
 
-pub fn create_type_map(input: &Path, output: &Path) {
+pub fn create_type_map(input: &Path, output: &Path, parallel: bool) {
     let buf_in = io::get_reader(input);
     let mut buf_out = io::get_writer(output);
     let mut triples = io::parse_ntriples(buf_in);
-    while !triples.is_end() {
-        triples
-            .parse_step(&mut |t| index_triple(t, &mut buf_out))
-            .unwrap();
+
```

Comment:
**comment:** on my side, this reduces performance compared to single thread. Maybe due to overhead of passing memory between cores?

**question:** should try use this pattern on pseudo instead? It is more likely to be CPU-bound

Reply:
jeah thats correct. I just started with index first, to see how it works.
 parallelization of index does not make sense as one can see by testing it.
 
 Intersting is however where the overhead lies: probably just due to passing to threads + locking the Buf constantly... Also the benefits on pseudo will be questional if it really works =). 
I think: Leave it as first good solution, come back to it later.
 
I am adding the parallelization on  pseudo.